pandas/_libs/tslibs/src/datetime/np_datetime_strings.c

```diff
@@ -104,19 +107,44 @@ int parse_iso_8601_datetime(const char *str, int len, int want_exc,
     while (sublen > 0 && isspace(*substr)) {
         ++substr;
         --sublen;
+        if (format_len < 1) {
```

Comment:
OK looking great. Now I think we've established a pretty clear pattern for a function. So with inputs of:
  * format string
  * format length remaining
  * characters to compare
  * whether exactness matters or not

I think you can now convert this back into the reusable function you always wanted. Something like:

```c
// This function will advance the pointer on format
// and decrement characters_remaining by n on success
// On failure will return -1 without incrementing
int compare_format(const char **format, int *characters_remaining, 
                 const char *compare_to, int n, const int exact) {
  if (*characters_remaining < n) {
    if (exact) {
      // TODO: in the future we should set a PyErr here 
      // to be very clear about what went wrong
      return -1;
    } else {
      // TODO: same return value in this function as
      // above branch, but stub out a future where
      // we have a better error message
      return -1;
    }
  }
  else {
    if (!strncmp(*format, compare_to, n)) {
      // TODO: PyErr to differentiate what went wrong
      return -1;
    } else {
      *format += n;
      *characters_remaining -= n;
      return 0;
    }
  }
}
```

Calling this like `compare_format(&format, &format_len, '%y', 2, check)` should do the mutation you are looking for and give us a better pattern to set explicit error messages about what went wrong. Ideally would do this along with this function, but I think in a follow up it would be better if we replaced the `goto` with a `-1` return value + setting a PyError, as that is a common practice in CPython development